### ISSUE TITLE
[onert] Introduce Conv2D train kernel

### ISSUE THREAD

**Issue by ragmani** [Samsung/ONE#11174]
## What

Let's introduce Conv2D kernel for training.

For #11389

Draft #11669

## Tasks
- [x] Bring Conv2D kernels from tensorflow v2.12.1
  - #11988
- [x] Apply Conv2D kernels
  - #11989
  - #11990
### Conv2D Forward

$$
\begin{align}
Z_{i,j,k}^{(m_i)} &= \sum\limits_{a} \sum\limits_{b} \sum\limits_{c} X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{(m_i)} * W_{a,b,c,k}  + b_k, \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$

$a$ : weight height
$b$ : weight width
$c$ : input channel
$k$ : output channel
$i$ : output height
$j$ : output width
$m_i$ : $i$ th batch

$S_i$ : Stride height
$S_j$ : Stride width 
$P_t$ : Pad top
$P_l$ : Pad left

#### Input
$$
\begin{align}
\frac{\partial L}{\partial X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{m_i}}&= \sum\limits_{k} \sum\limits_{a} \sum\limits_{b} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} W_{a,b,c,k} , \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$

#### Weights
$$
\begin{align}
\frac{\partial L}{\partial W_{a,b,c,k}}  &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{m_i} , \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$

#### Bias
$$
\begin{align}
\frac{\partial L}{\partial b_k} &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}}  \frac{\partial Z_{i,j,c}^{m_i}}{\partial b_k} \\
    &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} {\cdot } 1 \\
    &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}}
\end{align}
$$

**Comment by ragmani** [comment]
We decided to implement other operations first to train the model that does not include this operation as the first test model. So this task is postponed.

**Comment by jyoungyun** [comment]
I made an issue just to manage the remaining things. You can keep going on about this issue. :)
I will implement other things.

**Comment by ragmani** [comment]
How about joining this issue with #11389? If so, this issue will become a sub-issue of #11389

**Comment by ragmani** [comment]
![image](https://github.com/Samsung/ONE/assets/18050611/75af6ff9-eefe-4680-84b3-77760a9a2e38)

**Comment by ragmani** [comment]
- The result with the draft #11669 
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train mnist_keras_with_conv.circle --load_expected:raw out/train.output.1000.bin --load_input:raw out/train.input.1000.bin --data_length 1000 --learning_rate 0.001 --optimizer 1
Model Expected Filename out/train.output.1000.bin
Model Input Filename out/train.input.1000.bin
Model Filename mnist_keras_with_conv.circle
Epoch 1/5 - 4.316ms/step - loss: [0] 0.0617
Epoch 2/5 - 3.831ms/step - loss: [0] 0.0355
Epoch 3/5 - 4.085ms/step - loss: [0] 0.0271
Epoch 4/5 - 3.976ms/step - loss: [0] 0.0220
Epoch 5/5 - 4.231ms/step - loss: [0] 0.0183
===================================
MODEL_LOAD   takes 0.622 ms
PREPARE      takes 4.847 ms
EXECUTE      takes 637.568 ms
- MEAN     :  637.568 ms
- MAX      :  637.568 ms
- MIN      :  637.568 ms
- GEOMEAN  :  637.568 ms
===================================
```

**Comment by ragmani** [comment]
- Test model
![Screenshot from 2023-10-13 21-56-37](https://github.com/Samsung/ONE/assets/18050611/b0847e4c-18ff-4019-b14e-5eb39112911b)

- Result
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train --modelfile mnist_conv_trained.circle --load_input:raw models/mnist/train.input.1000.bin --load_expected:raw models/mnist/train.output.1000.bin --data_length 1000 --optimizer 1
Model Expected Filename models/mnist/train.output.1000.bin
Model Input Filename models/mnist/train.input.1000.bin
Model Filename mnist_conv_trained.circle
Epoch 1/5 - 156.727ms/step - loss: [0] 75.1892
Epoch 2/5 - 149.585ms/step - loss: [0] 65.3398
Epoch 3/5 - 177.503ms/step - loss: [0] 58.1592
Epoch 4/5 - 186.076ms/step - loss: [0] 52.1546
Epoch 5/5 - 190.967ms/step - loss: [0] 46.9015
===================================
MODEL_LOAD   takes 0.930 ms
PREPARE      takes 48.824 ms
EXECUTE      takes 26696.621 ms
- MEAN     :  26696.621 ms
- MAX      :  26696.621 ms
- MIN      :  26696.621 ms
- GEOMEAN  :  26696.621 ms
===================================
```

Training of the conv2D model does not work well. I guess there seems to be a problem with calculating the gradient of weights.

**Comment by ragmani** [comment]
It works well with the above model now.
```bash
$ ./Product/x86_64-linux.release/out/bin/onert_train mnist_conv_trained.circle --load_expected:raw out/train.output.1000.bin --load_input:raw out/train.input.1000.bin --data_length 1000 --learning_rate=0.0001 --optimizer 1 
Model Expected Filename out/train.output.1000.bin
Model Input Filename out/train.input.1000.bin
Model Filename mnist_conv_trained.circle
Epoch 1/5 - time: 116.608ms/step - loss: [0] 8.2901
Epoch 2/5 - time: 116.787ms/step - loss: [0] 0.4530
Epoch 3/5 - time: 116.664ms/step - loss: [0] 0.2124
Epoch 4/5 - time: 116.597ms/step - loss: [0] 0.1512
Epoch 5/5 - time: 116.355ms/step - loss: [0] 0.1216
===================================
MODEL_LOAD   takes 0.4820 ms
PREPARE      takes 16.5450 ms
EXECUTE      takes 18080.2070 ms
- Epoch 1      takes 3614.8600 ms
- Epoch 2      takes 3620.4110 ms
- Epoch 3      takes 3616.5890 ms
- Epoch 4      takes 3614.5210 ms
- Epoch 5      takes 3607.0000 ms
===================================
```

**Comment by ragmani** [comment]
Done.
